core/App/ehsm_napi.cpp

```diff
@@ -1139,83 +1140,549 @@ char* NAPI_ExportDataKey(const char* cmk_base64,
 }
 
 /*
-@return
-[string] json string
+ * process receive msg2 json string to sgx_ra_msg2_t
+ *  @param ra_msg2 : receive msg2 json string
+ *  @param req_msg2 : return sgx_ra_msg2_t
+ *  @param msg2_size : return sgx_ra_msg2_t real size
+ */
+ehsm_status_t ra_proc_msg2(std::string ra_msg2, sgx_ra_msg2_t *req_msg2,uint32_t *msg2_size)
+{
+    if(ra_msg2.empty() || req_msg2 == NULL || msg2_size == NULL)
     {
-        code: int,
-        message: string,
-        result: {
-            challenge_base64 : string,
-            ga_base64 : string
+        return EH_ARGUMENTS_BAD;
+    }
+
+    JsonObj msg2_json;
+    
+    if(msg2_json.parse(ra_msg2))
+    {
+        std::string json_key;
+        // g_b
+        json_key.clear();
+        json_key = json_key + "g_b" + LAYERED_CHARACTER + "gx";
+        msg2_json.readData_uint8Array(json_key, req_msg2->g_b.gx);
+        json_key.clear();
+        json_key = json_key + "g_b" + LAYERED_CHARACTER + "gy";
+        msg2_json.readData_uint8Array(json_key, req_msg2->g_b.gy);
+
+        // spid
+        json_key.clear();
+        json_key = json_key + "spid" + LAYERED_CHARACTER + "id";
+        msg2_json.readData_uint8Array(json_key, req_msg2->spid.id);
+
+        // quote_type
+        req_msg2->quote_type = msg2_json.readData_uint16("quote_type");
+
+        // kdf_id
+        req_msg2->kdf_id = msg2_json.readData_uint16("kdf_id");
+
+        // // sign_gb_ga
+        json_key.clear();
+        json_key = json_key + "sign_gb_ga" + LAYERED_CHARACTER + "x";
+        msg2_json.readData_uint32Array(json_key, req_msg2->sign_gb_ga.x);
+        json_key.clear();
+        json_key = json_key + "sign_gb_ga" + LAYERED_CHARACTER + "y";
+        msg2_json.readData_uint32Array(json_key, req_msg2->sign_gb_ga.y);
+
+        // mac
+        msg2_json.readData_uint8Array("mac",req_msg2->mac);
+
+        // sig_rl_size
+        req_msg2->sig_rl_size = msg2_json.readData_uint32("sig_rl_size");
+
+        // sig_rl
+        msg2_json.readData_uint8Array("sig_rl",req_msg2->sig_rl);
+
+        // setting return msg2_size
+        *msg2_size = sizeof(sgx_ra_msg2_t) + req_msg2->sig_rl_size;
+    }
+    else
+    {
+        return EH_ARGUMENTS_BAD;
+    }
+    return EH_OK;
+}
+
+/*
+ * process will be return to Enroll APP msg3 json
+ *  @param p_msg3 : will be parse sgx_ra_msg3_t
+ *  @param quote_size : quote size
+ *  @param retJsonObj : return msg3 json object 
+ *  retJsonObj : RetJsonObj
+ *      {
+ *          "mac" : array(int),
+ *          "g_a" : Json::Value
+ *              {
+ *                  gx : array(int),
+ *                  gy : array(int)
+ *              },
+ *          "ps_sec_prop" : Json::Value
+ *              {
+ *                  sgx_ps_sec_prop_desc : array(int)
+ *              },
+ *          "quote_size" : int,
+ *          "quote" : array(int)
+ *      }
+ */
+ehsm_status_t ra_proc_msg3(sgx_ra_msg3_t *p_msg3, RetJsonObj *retJsonObj, uint32_t quote_size)
+{
```

Comment:
static

Reply:
marshal_msg3_to_json

unmarshal_msg3_from_json

Reply:
The function will be move to marshal.cpp and changed name.